dll/win32/shlwapi/propbag.cpp

```diff
@@ -0,0 +1,318 @@
+ * LICENSE:     LGPL-2.0-or-later (https://spdx.org/licenses/LGPL-2.0-or-later)
+ * PURPOSE:     Implement shell property bags
+ * COPYRIGHT:   Copyright 2023 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
+ */
+
+#include "precomp.h"
+#include <shlwapi.h>
+#include <shlwapi_undoc.h>
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+// Dynamic structure array (DSA); See COMCTL32
+struct _DSA
+{
+    INT  nItemCount;
+    LPVOID pData;
+    INT  nMaxCount;
+    INT  nItemSize;
+    INT  nGrow;
+};
+
+// Property Bag Entry
+typedef struct tagPBAGENTRY
+{
+    LPWSTR pszPropName;
+    VARIANT vari;
+} PBAGENTRY, *PPBAGENTRY;
+
+class CBasePropertyBag : public IPropertyBag, public IPropertyBag2
+{
+protected:
+    LONG m_cRefs;   // reference count
+    DWORD m_dwMode; // STGM_* flags
+
+public:
+    CBasePropertyBag(DWORD dwMode)
+        : m_cRefs(1)
+        , m_dwMode(dwMode)
+    {
+    }
+    virtual ~CBasePropertyBag()
+    {
+    }
+
+    // IUnknown interface
+    STDMETHODIMP QueryInterface(REFIID riid, void **ppvObject) override
+    {
+        if (::IsEqualGUID(riid, IID_IPropertyBag2))
+        {
+            AddRef();
+            *ppvObject = dynamic_cast<IPropertyBag2*>(this);
+            return S_OK;
+        }
+        if (::IsEqualGUID(riid, IID_IUnknown) || ::IsEqualGUID(riid, IID_IPropertyBag))
+        {
+            AddRef();
+            *ppvObject = dynamic_cast<IPropertyBag*>(this);
+            return S_OK;
+        }
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP_(ULONG) AddRef() override
+    {
+        return ::InterlockedIncrement(&m_cRefs);
+    }
+    STDMETHODIMP_(ULONG) Release() override
+    {
+        if (::InterlockedDecrement(&m_cRefs) == 0)
+        {
+            delete this;
+            return 0;
+        }
+        return m_cRefs;
+    }
+
+    // IPropertyBag interface (pure)
+    STDMETHODIMP Read(LPCWSTR pszPropName, VARIANT *pvari, IErrorLog *pErrorLog) override = 0;
+    STDMETHODIMP Write(LPCWSTR pszPropName, VARIANT *pvari) override = 0;
+
+    // IPropertyBag2 interface (stubs)
+    STDMETHODIMP Read(
+        ULONG cProperties,
+        PROPBAG2 *pPropBag,
+        IErrorLog *pErrorLog,
+        VARIANT *pvarValue,
+        HRESULT *phrError) override
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP Write(
+        ULONG cProperties,
+        PROPBAG2 *pPropBag,
+        VARIANT *pvarValue)
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP CountProperties(ULONG *pcProperties) override
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP GetPropertyInfo(
+        ULONG iProperty,
+        ULONG cProperties,
+        PROPBAG2 *pPropBag,
+        ULONG *pcProperties) override
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP LoadObject(
+        LPCWSTR pstrName,
+        DWORD dwHint,
+        IUnknown *pUnkObject,
+        IErrorLog *pErrorLog) override
+    {
+        return E_NOTIMPL;
+    }
+
+    static void* operator new(size_t size)
```

Comment:
Is there a reason that you need to overload this?

Reply:
We cannot use both `<new>` and `nothrow` yet.

Reply:
Deleted.